src/components/ContactList.tsx

```diff
@@ -37,20 +47,66 @@ const ContactList = () => {
         fetchContacts();
     }, []) 
 
+    //TODO: optimize search so that it can find partial match to name search
+    //TODO: fix the any typing in matchByNameSearResults
+    const getMatchingNames = (nameStr: string | null) => {
+        const namesArray: any[] = [];
+        contacts.map((contact: any) => {
+            if (contact.name === nameStr) {
+                namesArray.push(contact.name);
+            }
+        })
+        return namesArray;
+    }
+
+    // console.log("search results: ", getMatchingNames(query))
+
+    //TODO: if I want to have a search for all values nested in contact, then I can start here
+    // let contactValues: unknown[] = [];
+    // if (contacts.length) {
+    //     console.log("Line 51: ", contacts);
+    //     contactValues = contacts.flatMap(contact  => {
+    //         const allNestedValuesInObject = [];
+    //         for (const property in contact) {
+    //             if (typeof contact[property] === 'object' && contact[property] !== null) {
+    //                 allNestedValuesInObject.push(Object.values(contact[property]));
+    //             } else {
+    //                 allNestedValuesInObject.push(contact[property]);
+    //             }
+    //         }
+    //         return allNestedValuesInObject.flat();
+    //     })
+    // }
+
     
-    //TODO: fix TS errors
     return contacts.length > 0 ? (
         <>
             <ul>
                 {contacts.map(contact => {
-                    return (
-                        <ContactListItem name={(contact as ContactListItemType).name} id={(contact as ContactListItemType).id} />
-                    );
-                })}
+                    console.log("query value is: ", query);
+                    console.log("Truthiness of query: ", new Boolean(query).valueOf());
+                    if (!query) {
+                        return <ContactListItem name={(contact as ContactListItemType).name} id={(contact as ContactListItemType).id} />
+                    }
+                    if (getMatchingNames(query).includes((contact as ContactListItemType).name)) {
+                        return <ContactListItem name={(contact as ContactListItemType).name} id={(contact as ContactListItemType).id} />
+                    }
```

Comment:
Instead of doing this includes call, which is going to create O(n^2) performance for contacts length * matching names length - maybe you could create a separate filteredContacts variable and map over that, instead of using map as a filter as you are here?

Reply:
Sure, I can try that. General question: is it usually better to reduce the time complexity of your code even if it means increasing the space complexity with added variables?

Reply:
To answer that, in short, it depends. But as a rule of thumb O(n) space and O(n) time still beats the heck out of O(n^2) time.